tidb-lightning: add 50 TB data import best practices (#14271)
pingcap/docs-cn: tidb-lightning/data-import-best-practices.md

---
title: 50 TiB 数据导入最佳实践

> summary: 根据导入大单表的经验总结出的导入大量数据的一套最佳实践。

Reviewer: ```suggestion
summary: 了解将大规模数据导入 TiDB 的最佳实践。
```

- 执行计划

    Checksum 通过后，会对每个表执行 [`ANALYZE TABLE`](/sql-statements/sql-statement-analyze-table.md)，构建最佳的执行计划。当有很多表或单个表很大时，ANALYZE 阶段耗时会很长。

- 相关 Issue

    在实际导入 50 TiB 数据的过程中，存在一些在海量源文件及大规模集群下才会暴露出的问题。在选择产品版本时，请检查该版本是否已经修复了某些影响导入性能的 Issue。以下 Issue 在 v6.5.3、v7.1.0 及更新的版本中都已修复：

    - [Issue-14745](https://github.com/tikv/tikv/issues/14745)：导入完成后，TiKV Import 目录遗留大量临时文件。
    - [Issue-6426](https://github.com/tikv/pd/issues/6426)：PD [范围调度](/tidb-lightning/tidb-lightning-physical-import-mode-usage.md#导入时暂停-pd-调度的范围)接口存在未打散 Region 的情况，导致 Scatter Region 超时。v6.2.0 之前采用停止全局调度的方式，不会出现该问题。
    - [Issue-43079](https://github.com/pingcap/tidb/pull/43079)：TiDB Lightning 对 NotLeader 错误的重试未刷新 Region Peers 信息。
    - [Issue-43291](https://github.com/pingcap/tidb/issues/43291)：TiDB Lightning 未对临时文件删除的情况（"No such file or directory"）进行重试。

## 准备源文件

- 在生成文件时，在单个文件内尽量按照主键排序；如果表定义没有主键，可以添加一个自增主键，此时对文件内容顺序无要求。
- 在给多个 TiDB Lightning 实例分配要导入的源文件时，尽量避免多个源文件之间存在重叠的主键或非空唯一索引的情况。如果生成文件是全局有序，可以按照范围划分不同的文件给不同 TiDB Lightning 实例进行导入，达到最佳导入效果。
- 在生成文件时，每个文件尽量控制在 96 MiB 以下。

> - 如果文件特别大，超过 256 MiB，需要开启 [strict-format](/migrate-from-csv-files-to-tidb.md#第-4-步导入性能优化可选)。

Reviewer: ```suggestion
- 如果文件特别大，超过 256 MiB，需要开启 [`strict-format`](/migrate-from-csv-files-to-tidb.md#第-4-步导入性能优化可选)。
```

- `disk-quota`：尽量保证 TiDB Lightning 排序目录空间大于数据源大小。如无法保证，可以设置 `disk-quota` 为 TiDB Lightning 排序目录空间的 80%。此时 TiDB Lightning 会按照 `disk-quota` 的大小为一个批次去排序、写入，导入性能低于完整排序。
- `GOMEMLIMIT`：TiDB Lightning 采用 Go 语言实现，设置 `GOMEMLIMIT` 为实例内存的 80%，降低因为 Go GC 机制带来的 OOM 概率。

关于 TiDB Lightning 参数信息，请参考 [TiDB Lightning 配置参数](/tidb-lightning/tidb-lightning-configuration.md)。

## 解决 "checksum mismatch" 问题

在数据校验过程中，可能会出现冲突数据，报错信息为："checksum mismatch"。出现该问题，可以按照以下思路解决：

1. 排查源数据是否有主键、唯一键冲突，解决冲突后再重新导入。根据以往经验，主键和唯一键冲突是导致该报错的主要原因。
2. 表的主键、唯一键定义是否合理。如果不合理，更改表定义后重新导入。
3. 如果按上述两个步骤操作后仍未解决问题，需要进一步判断源数据中是否有少量（低于 10%）不可预期的冲突数据。如果需要 TiDB Lightning 帮助检测、解决冲突数据，需要开启[冲突检测功能](/tidb-lightning/tidb-lightning-physical-import-mode-usage.md#冲突数据检测)。

## 开启断点续传

大规模的数据导入，一定要参照[断点续传](/tidb-lightning/tidb-lightning-checkpoints.md)文档，开启断点续传，并推荐优先使用 MySQL 作为 Driver，避免因为 TiDB Lightning 运行在容器环境，容器退出后断点信息被一并删除。

如果导入过程中遇到下游 TiKV 空间不足，可以手动执行 `kill` 命令关闭（不要带 `-9` 选项）所有 TiDB Lightning 实例，待扩容后，基于断点信息继续导入。

## 导入大单表的最佳实践

多表导入会导致 Checksum、ANALYZE 时间的增加，甚至超过数据导入本身，但是一般不需要调整配置。如果多表中存在单个或多个大表的情况，可以把这类大表的源文件划分出来，单独进行导入。

本小节重点介绍大单表导入的最佳实践。大单表没有严格的定义，一般认为符合以下任一条件者即为大单表：

- 大小超过 10 TiB
- 行数超过 10 亿、列数超过 50 的宽表

> ### 准备源文件

Reviewer: 同一篇文档里尽量避免相同的文内标题，以免在引用到具体小节时造成困惑。
```suggestion
### 生成源文件
```

> 根据上述源文件准备的步骤产生源文件，对于大单表，如果不能做到全局有序，但是可以做到文件内按主键有序，且是标准的 CSV 文件，可以尽量生成单个大文件（例如每个 20 GiB），然后开启 [`strict-format`](/migrate-from-csv-files-to-tidb.md#第-4-步导入性能优化可选)，既可以降低 TiDB Lightning 实例之间导入的数据文件中存在主键和唯一键的重叠，又能在导入前由 TiDB Lightning 实例对大文件进行切分，达到最佳的导入速度。

Reviewer: ```suggestion
根据上述[准备源文件的步骤](#准备源文件)产生源文件，对于大单表，如果不能做到全局有序，但是可以做到文件内按主键有序，且是标准的 CSV 文件，可以尽量生成单个大文件（例如每个 20 GiB），然后开启 [`strict-format`](/migrate-from-csv-files-to-tidb.md#第-4-步导入性能优化可选)，既可以降低 TiDB Lightning 实例之间导入的数据文件中存在主键和唯一键的重叠，又能在导入前由 TiDB Lightning 实例对大文件进行切分，达到最佳的导入速度。
```
